### ISSUE TITLE
Filter functions such as filter_has_var() accept an integer or a predefined constant (INPUT_*) as the first argument

### ISSUE THREAD

**Issue by davidg8080** [php/doc-en#3886]
From manual page: https://php.net/filter.constants

--- 

The documentation does not make reference to the integers which reference the different predefined constants. 

Here's a list of valid `INPUT_*` constants and their corresponding integer values:

- `INPUT_GET` corresponds to `1` (for accessing `$_GET` variables).
- `INPUT_POST` corresponds to `2` (for accessing `$_POST` variables).
- `INPUT_COOKIE` corresponds to `3` (for accessing `$_COOKIE` variables).
- `INPUT_ENV` corresponds to `4` (for accessing environment variables).
- `INPUT_SERVER` corresponds to `5` (for accessing `$_SERVER` variables).
- `INPUT_SESSION` corresponds to `6` (for accessing `$_SESSION` variables). This is available starting from PHP 7.0.0.
- `INPUT_REQUEST` corresponds to `99` (for accessing `$_REQUEST` variables). However, this constant is less commonly used in practice.

In the same way filter flag options in functions such as filter_input() expects an integer according to the documentation:

filter_input(
    int $type,
    string $var_name,
    int $filter = FILTER_DEFAULT,
    array|int $options = 0
): mixed

A list of sanitization and validation filter as strings, as well as flags as strings, are provided but no reference is ever made as to the numbers that the correspond to these. 

### **General Filters:**

- `FILTER_DEFAULT` (used when no specific filter is specified): **516**
- `FILTER_VALIDATE_BOOLEAN`: **258**
- `FILTER_VALIDATE_EMAIL`: **274**
- `FILTER_VALIDATE_FLOAT`: **259**
- `FILTER_VALIDATE_INT`: **257**
- `FILTER_VALIDATE_IP`: **275**
- `FILTER_VALIDATE_REGEXP`: **272**
- `FILTER_VALIDATE_URL`: **273**
- `FILTER_VALIDATE_DOMAIN`: **277**
- `FILTER_VALIDATE_MAC`: **276**

### **Sanitization Filters:**

- `FILTER_SANITIZE_EMAIL`: **517**
- `FILTER_SANITIZE_ENCODED`: **514**
- `FILTER_SANITIZE_NUMBER_FLOAT`: **520**
- `FILTER_SANITIZE_NUMBER_INT`: **519**
- `FILTER_SANITIZE_SPECIAL_CHARS`: **515**
- `FILTER_SANITIZE_FULL_SPECIAL_CHARS`: **522**
- `FILTER_SANITIZE_STRING` (deprecated as of PHP 8.1.0): **513**
- `FILTER_SANITIZE_URL`: **518**
- `FILTER_SANITIZE_ADD_SLASHES`: **523**

### **Other Filters:**

- `FILTER_UNSAFE_RAW`: **516** (this is the same as `FILTER_DEFAULT`, meaning no filtering will be applied).

#### **Flags for `FILTER_VALIDATE_INT`:**

- `FILTER_FLAG_ALLOW_OCTAL`: **1**
  - Allows octal numbers (e.g., `0123`).
- `FILTER_FLAG_ALLOW_HEX`: **2**
  - Allows hexadecimal numbers (e.g., `0x1A`).

#### **Flags for `FILTER_VALIDATE_BOOLEAN`:**

- `FILTER_NULL_ON_FAILURE`: **134217728**
  - Returns `NULL` for unrecognized values, instead of `FALSE`.

#### **Flags for `FILTER_VALIDATE_URL`:**

- `FILTER_FLAG_SCHEME_REQUIRED`: **1**
  - Requires the URL to include a scheme (e.g., `http`, `https`).
- `FILTER_FLAG_HOST_REQUIRED`: **2**
  - Requires the URL to include a host name (e.g., `www.example.com`).
- `FILTER_FLAG_PATH_REQUIRED`: **4**
  - Requires the URL to include a path (e.g., `/somepage`).
- `FILTER_FLAG_QUERY_REQUIRED`: **8**
  - Requires the URL to include a query string (e.g., `?id=123`).

#### **Flags for `FILTER_VALIDATE_EMAIL`:**

- `FILTER_FLAG_EMAIL_UNICODE`: **1048576**
  - Allows the email address to contain Unicode characters.

#### **Flags for `FILTER_VALIDATE_IP`:**

- `FILTER_FLAG_IPV4`: **1048576**
  - Validates only IPv4 addresses.
- `FILTER_FLAG_IPV6`: **2097152**
  - Validates only IPv6 addresses.
- `FILTER_FLAG_NO_PRIV_RANGE`: **8388608**
  - Fails if the IP address is within a private range (e.g., `192.168.x.x`).
- `FILTER_FLAG_NO_RES_RANGE`: **4194304**
  - Fails if the IP address is within a reserved range.

#### **Flags for `FILTER_SANITIZE_STRING` (deprecated in PHP 8.1):**

- `FILTER_FLAG_NO_ENCODE_QUOTES`: **2**
  - Does not encode single and double quotes.
- `FILTER_FLAG_STRIP_LOW`: **4**
  - Strips characters with ASCII values less than 32.
- `FILTER_FLAG_STRIP_HIGH`: **8**
  - Strips characters with ASCII values greater than 127.

#### **Flags for `FILTER_SANITIZE_NUMBER_FLOAT`:**

- `FILTER_FLAG_ALLOW_FRACTION`: **4096**
  - Allows a fractional part in the number (e.g., `123.45`).
- `FILTER_FLAG_ALLOW_THOUSAND`: **8192**
  - Allows thousand separators (e.g., `1,000`).
- `FILTER_FLAG_ALLOW_SCIENTIFIC`: **16384**
  - Allows scientific notation (e.g., `1e10`).

#### **Flags for `FILTER_SANITIZE_SPECIAL_CHARS` and `FILTER_SANITIZE_FULL_SPECIAL_CHARS`:**

- `FILTER_FLAG_NO_ENCODE_QUOTES`: **2**
  - Does not encode single and double quotes.

#### **Flags for `FILTER_SANITIZE_ENCODED`:**

- `FILTER_FLAG_STRIP_LOW`: **4**
  - Strips characters with ASCII values less than 32.
- `FILTER_FLAG_STRIP_HIGH`: **8**
  - Strips characters with ASCII values greater than 127.
- `FILTER_FLAG_ENCODE_LOW`: **16**
  - Encodes characters with ASCII values less than 32.
- `FILTER_FLAG_ENCODE_HIGH`: **32**
  - Encodes characters with ASCII values greater than 127.

### **Miscellaneous Flags:**

- `FILTER_FLAG_EMPTY_STRING_NULL`: **256**
  - Treats an empty string as `NULL`.

**Comment by damianwadley** [comment]
> A list of sanitization and validation filter as strings, as well as flags as strings, are provided but no reference is ever made as to the numbers that the correspond to these.

That is intentional. Putting aside how it's universally bad practice to use magic numbers in code, one of the main reasons constants exist is to protect yourself against the exact values of those constants ever changing. And PHP may very well change those values at any time.

If anyone wants to know the numbers backing those constants then they're free to do exactly what you did. But the constants are the correct "values" to be using in code.

**Comment by davidg8080** [comment]
I cannot understand how it would be considered bad practice because the documentation for these particular functions are written and documented expecting an integer. The user has no way to know what the "magic" is behind the scenes. I am not making a call to use the magic numbers (of which the documentation is unavailable), but I am suggesting the revision of the documentation for the specific page I listed. 

As an example the filter_input() function is depicted as so: 

```
filter_input(
**int** $type,
string $var_name,
**int** $filter = FILTER_DEFAULT,
array|**int** $options = 0
): mixed
```

After reviewing the function one would expect the Parameters: type to be a list of integers but instead a list of predefined constants without any reference to numeric (int) values is listed. 

Thee function reads **one of** and then the predefined constants. This applies to the pages for the following filters: 

- [filter_has_var()](https://www.php.net/manual/en/function.filter-has-var.php)
- [filter_input()](https://www.php.net/manual/en/function.filter-input.php) 
- [filter_input_array](https://www.php.net/manual/en/function.filter-input-array.php)

In the documentation for these **_very useful_** functions, the filter type is listed as expecting an integer (int) and no reference is made to the integer but instead an incomplete list of predefined constants is presented see that **INPUT_SESSION** and **INPUT_REQUEST** are not included but have been available as of php 5.2 and 5.4 I understand. After this by clicking on any of the types listed on these functions brings you to the [page referenced](https://www.php.net/manual/en/filter.constants.php). 

Listed predefined constants appear incorrectly on the filter constants page stating that the below:

- INPUT_SESSION (not implemented yet) 
- INPUT_REQUEST (not implemented yet) 

In general the same case applies with the $filter argument and $filter flags in terms of non documented numbers that relate to the (int) expected by the listed functions and the other functions that accept validation or sanitization flags.  

Before you close the issue you could consider clarifying the use of untrustworthy _"magic numbers"_  and provide the user with updated information for these php functions. The documentation about these magic numbers that equal a seemingly arbitrary predefined constants and filter flags. Regardless of the reasoning, I understand you don't have a type cast for predefined constants that exist but not documenting the fact that the function expects and runs on undocumented integers means the documentation should be revised to address the central issue and the other issues Ive listed. 

I ask this be considered for the benefit of the community. Thank you

**Comment by cmb69** [comment]
A "magic number" in programming refers to any literal number (maybe exempt zero and one) you use in code. That has nothing to do with ext/filter, but is generally considered bad practise (or at least questionable).

While I agree that the documentation of the filter extension could be improved (like almost all of the PHP manual), I also agree with @damianwadley that we do not want to document the values of constants (maybe except for error constants). Those who are interested in these numbers can look them up in the source code, and use them at their own risk.

**Comment by davidg8080** [comment]
> A "magic number" in programming refers to any literal number (maybe exempt zero and one) you use in code. That has nothing to do with ext/filter, but is generally considered bad practise (or at least questionable).
> 
> While I agree that the documentation of the filter extension could be improved (like almost all of the PHP manual), I also agree with @damianwadley that we do not want to document the values of constants (maybe except for error constants). Those who are interested in these numbers can look them up in the source code, and use them at their own risk.

Its only considered bad practice because the development team will change the numbers at will and they may very well have a valid reason to do so.  However, we don't have benchmarks as to the speed of using the corresponding integer is faster than looking up the predefined constant. If there is a performance advantage it would be perfectly reasonable to initiate the $filter_type method as such: 

```
switch ($input_method) {
  case 'GET':     $input_method_num =  1;  break; // (int) 01 = predefined constant INPUT_GET
  case 'POST':    $input_method_num =  2;  break; // (int) 02 = predefined constant INPUT_POST
  case 'COOKIE':  $input_method_num =  3;  break; // (int) 03 = predefined constant INPUT_COOKIE
  case 'SERVER':  $input_method_num =  4;  break; // (int) 04 = predefined constant INPUT_SERVER
  case 'ENV':     $input_method_num =  5;  break; // (int) 05 = predefined constant INPUT_ENV
  case 'SESSION': $input_method_num =  6;  break; // (int) 06 = predefined constant INPUT_SESSION
  case 'REQUEST': $input_method_num = 99;  break; // (int) 99 = predefined constant INPUT_REQUEST
  default:
    die('response method not valid'); 
}
```

But Im not disputing this since as @damianwadley these numbers change. What I am challenging is the way that the functions are written to explicitly expect an (int) but then the documentation goes on to give an **_incomplete_** list of **STRINGS** known as predefined constants. No mention as to why the function is written as filter_input(**(int)** $type, ...) and the same applies to the $flags argument which also expects an integer.  The [page for the function filter_list()](https://www.php.net/manual/en/function.filter-list.php) the example section shows the undocumented changes in following lists: 

```
Array
(
    [0] => int
    [1] => boolean
    [2] => float
    [3] => validate_regexp
    [4] => validate_url
    [5] => validate_email
    [6] => validate_ip
    [7] => string
    [8] => stripped
    [9] => encoded
    [10] => special_chars
    [11] => unsafe_raw
    [12] => email
    [13] => url
    [14] => number_int
    [15] => number_float
    [16] => magic_quotes
    [17] => callback
)
```

### PHP v7.00 filter_list() :  
```
Array
(
    [0] => int
    [1] => boolean
    [2] => float
    [3] => validate_regexp
    [4] => validate_domain
    [5] => validate_url
    [6] => validate_email
    [7] => validate_ip
    [8] => validate_mac
    [9] => string
    [10] => stripped
    [11] => encoded
    [12] => special_chars
    [13] => full_special_chars
    [14] => unsafe_raw
    [15] => email
    [16] => url
    [17] => number_int
    [18] => number_float
    [19] => magic_quotes
    [20] => callback
)
```
### PHP v8.00 filter_list() :  
```
Array
(
    [0] => int
    [1] => boolean
    [2] => float
    [3] => validate_regexp
    [4] => validate_domain
    [5] => validate_url
    [6] => validate_email
    [7] => validate_ip
    [8] => validate_mac
    [9] => string
    [10] => stripped
    [11] => encoded
    [12] => special_chars
    [13] => full_special_chars
    [14] => unsafe_raw
    [15] => email
    [16] => url
    [17] => number_int
    [18] => number_float
    [19] => add_slashes
    [20] => callback
)
```

There is no disagreement in what the best practices are according to the PHP development team. That is evident in the changes made in the changes to the magic numbers that reference the filters_list(), but that is not an excuse for the documentation should be so blatantly outdated. 

I believe that the correct use of the filter_input and filter_var set of functions is very important as they **CAN BE** safer and faster than htmlspecialchars() and strip_tags() but they are poorly documented. One must look elsewhere for the best use of the functions.

**Comment by cmb69** [comment]
> Its only considered bad practice because the development team will change the numbers at will

No, it's considered bad practise because a reader will not understand what the number means. You can add a comment, but why not use the constant right away?

> However, we don't have benchmarks as to the speed of using the corresponding integer is faster than looking up the predefined constant.

Well, if that is concerns you, better not use PHP at all.

> The [page for the function filter_list()](https://www.php.net/manual/en/function.filter-list.php) the example section shows the undocumented changes in following lists:

First, it states: "The above example will output something *similar* to:" (emphasis mine)

Second, I don't know why there even are functions like `filter_list()` and `filter_id()`. I guess that the filter extension has (or had) an extension mechanism where additional filters could be registered by other extensions; I'm not aware that this has been done in practise. In my opinion, it would be better to remove these functions, than to document the values of the filters.

**Comment by davidg8080** [comment]
> > Its only considered bad practice because the development team will change the numbers at will
> 
> No, it's considered bad practise because a reader will not understand what the number means. You can add a comment, but why not use the constant right away?

Thats perfect, because I'm advocating that the documentation be complete and written correctly. The function expects an (int) but we're supposed to know that integer is a "magic number" and we're really supposed to use a string. 2 of which are not listed as working; **INPUT_SESSION**  appears as (not implemented yet) and **INPUT_REQUEST** (not implemented yet) as well. 

> > However, we don't have benchmarks as to the speed of using the corresponding integer is faster than looking up the predefined constant.
> 
> Well, if that is concerns you, better not use PHP at all.

I fail to see the point. we're in the php documentation why would you compare to ANOTHER DIFFERENT language, btw I did the benchmarks and the difference is negligible at best. Sometimes using the (int) is faster and sometimes using the predefined constant is faster. Thats fine I never made a point about speed, I said UNDOCUMENTED STRINGS (predefined constants) are best practice but they wrote the functions expressing that the function expects an integer. This really isnt that hard. 
> 
> > The [page for the function filter_list()](https://www.php.net/manual/en/function.filter-list.php) the example section shows the undocumented changes in following lists:
> 
> First, it states: "The above example will output something _similar_ to:" (emphasis mine)
> 
> Second, I don't know why there even are functions like `filter_list()` and `filter_id()`. I guess that the filter extension has (or had) an extension mechanism where additional filters could be registered by other extensions; I'm not aware that this has been done in practise. In my opinion, it would be better to remove these functions, than to document the values of the filters.

Perfect, so finally we agree that the confusion arises as magic numbers are used interchangeably in different versions of PHP, numbers that are difficult to get but heres what Id like you to understand: If you use those functions that you would dispose of, you will confirm that 2 filter types have been implemented in PHP and are UNDOCUMENTED. 

I believe the issue is actually simpler. Ever since the php documentation changed to include type casting and type setting, the php documentation was adapted to express what type of argument the functions exist. In this case we see underneath the hood :that the function uses an integer to work and that integer is equal to a magic number that equals a predefined constant. However there is no type cast for predefined constants and therefore the function prints that the filter type and filter flags expected are integers but that developers really want you to use predefined constants. 

Lastly I did not ask for the functions documentation to be updated, I asked that the [filter.constants](https://php.net/filter.constants) page de updated specifically with the filters and flags which are incomplete. Why was this closed so quickly? The filter.constants page is incomplete and the page should explain best practices are not to use integers but use the predefined constants as those will likely not change in the future. 

I don't want to sound discourteous. I simply don't understand having to defend correct and complete documentation when the docs are clearly incomplete and can lead to confusion.

**Comment by Girgias** [comment]
Because we, from the documentation project, decided to *stop* documenting the value of constants when one should just use the constant directly.

Should these integers actually be an enum nowadays? Yes. However, this is an old extension.
Documenting the values of integers that can change at any given version, especially if it exposed from an external C library (which is not the case here but the core reasoning) is extremely impractical from the maintenance PoV.

So this is not a bug by the PHP documentation standards.

**Comment by davidg8080** [comment]
> Because we, from the documentation project, decided to _stop_ documenting the value of constants when one should just use the constant directly.
> 
> Should these integers actually be an enum nowadays? Yes. However, this is an old extension. Documenting the values of integers that can change at any given version, especially if it exposed from an external C library (which is not the case here but the core reasoning) is extremely impractical from the maintenance PoV.
> 
> So this is not a bug by the PHP documentation standards.

Girgias, Ill let this be my final comment of the subject and hopefully the team could consider the request. 

A. I am suggesting that the filters.constants page be updated as the filter types, the validation and sanitization flags are outdated. 

B. As said before it is confusing (or consider it perhaps a deficiency of the documentation) when the function(s) that lead to the filters.constants documentation page are explicitly written to expect an integer and no reference of integers is made. A simple paragraph would explain why the function was written that way. Ive been explained 7 times why we must use the predefined constants. 

C. Please momentarily put the integer/enum dichotomy aside for the purpose of this suggestion and check how many filter flags are missing from the documentation. Lastly I submit to your team: is there a reason that the **INPUT_SESSION** listed as _(not implemented yet)_ and **INPUT_REQUEST** _(not implemented yet)_ as well? 

Thank you and the team for your time.

**Comment by cmb69** [comment]
> I am suggesting that the filters.constants page be updated as the filter types, the validation and sanitization flags are outdated.

Okay, this is a valid request. FILTER_SANITIZE_FULL_SPECIAL_CHARS` and `FILTER_FLAG_GLOBAL_RANGE` are not documented there. What else would be outdated?

> is there a reason that the **INPUT_SESSION** listed as _(not implemented yet)_ and **INPUT_REQUEST** _(not implemented yet)_ as well?

These are actually removed as of PHP 8.0.0, so that needs to be updated. Generally, though, constants which are defined by PHP need to be documented, even if they have no effect whatsoever.

> As said before it is confusing (or consider it perhaps a deficiency of the documentation) when the function(s) that lead to the filters.constants documentation page are explicitly written to expect an integer and no reference of integers is made. A simple paragraph would explain why the function was written that way.

This is not particularly special for ext/filter. A lot of PHP functions expect integers, but are supposed to get a constant or sometimes a disjunction of constants. Of course, you can always pass the respective integer value instead archiving the same result. The function wouldn't know how exactly it has been called. You can even pass some expression which evaluates to an integer.

> Second, I don't know why there even are functions like `filter_list()` and `filter_id()`. I guess that the filter extension has (or had) an extension mechanism where additional filters could be registered by other extensions; […]

I have checked that, and at least as of PHP 5.3 there is no extension mechanism; the list of supported files is a hard-coded const array. So `filter_list()` and `filter_id()` don't really make sense, and should likely be removed after deprecation. That would, of course, require the RFC process, and isn't a documentation issue anyway.

**Comment by davidg8080** [comment]
> See #3920.

Thank you! Because we talked so much about the subject I found an error (bug) mistake in message. This is bad code but it I was able to trigger an error that references the constant number instead of the constant name. I understand why the error is displayed but before these messages wouldn't have been able to figure out what was going on: 

```
$x = constant('INPUT_POST'); // correct 
$y = constant(INPUT_POST); // Fatal error: Uncaught Error: Undefined constant "0" 
$z = constant(INPUT_GET); // Fatal error: Uncaught Error: Undefined constant "1" 
```

So the `get_defined_constants(true)` function displays a multi-dimensional array with all of the predefined constants we would have use of available in PHP 8.3.12
```
    [filter] => Array
        (
            [INPUT_POST] => 0
            [INPUT_GET] => 1
            [INPUT_COOKIE] => 2
            [INPUT_ENV] => 4
            [INPUT_SERVER] => 5
            [FILTER_FLAG_NONE] => 0
            [FILTER_REQUIRE_SCALAR] => 33554432
            [FILTER_REQUIRE_ARRAY] => 16777216
            [FILTER_FORCE_ARRAY] => 67108864
            [FILTER_NULL_ON_FAILURE] => 134217728
            [FILTER_VALIDATE_INT] => 257
            [FILTER_VALIDATE_BOOLEAN] => 258
            [FILTER_VALIDATE_BOOL] => 258
            [FILTER_VALIDATE_FLOAT] => 259
            [FILTER_VALIDATE_REGEXP] => 272
            [FILTER_VALIDATE_DOMAIN] => 277
            [FILTER_VALIDATE_URL] => 273
            [FILTER_VALIDATE_EMAIL] => 274
            [FILTER_VALIDATE_IP] => 275
            [FILTER_VALIDATE_MAC] => 276
            [FILTER_DEFAULT] => 516
            [FILTER_UNSAFE_RAW] => 516
            [FILTER_SANITIZE_STRING] => 513
            [FILTER_SANITIZE_STRIPPED] => 513
            [FILTER_SANITIZE_ENCODED] => 514
            [FILTER_SANITIZE_SPECIAL_CHARS] => 515
            [FILTER_SANITIZE_FULL_SPECIAL_CHARS] => 522
            [FILTER_SANITIZE_EMAIL] => 517
            [FILTER_SANITIZE_URL] => 518
            [FILTER_SANITIZE_NUMBER_INT] => 519
            [FILTER_SANITIZE_NUMBER_FLOAT] => 520
            [FILTER_SANITIZE_ADD_SLASHES] => 523
            [FILTER_CALLBACK] => 1024
            [FILTER_FLAG_ALLOW_OCTAL] => 1
            [FILTER_FLAG_ALLOW_HEX] => 2
            [FILTER_FLAG_STRIP_LOW] => 4
            [FILTER_FLAG_STRIP_HIGH] => 8
            [FILTER_FLAG_STRIP_BACKTICK] => 512
            [FILTER_FLAG_ENCODE_LOW] => 16
            [FILTER_FLAG_ENCODE_HIGH] => 32
            [FILTER_FLAG_ENCODE_AMP] => 64
            [FILTER_FLAG_NO_ENCODE_QUOTES] => 128
            [FILTER_FLAG_EMPTY_STRING_NULL] => 256
            [FILTER_FLAG_ALLOW_FRACTION] => 4096
            [FILTER_FLAG_ALLOW_THOUSAND] => 8192
            [FILTER_FLAG_ALLOW_SCIENTIFIC] => 16384
            [FILTER_FLAG_PATH_REQUIRED] => 262144
            [FILTER_FLAG_QUERY_REQUIRED] => 524288
            [FILTER_FLAG_IPV4] => 1048576
            [FILTER_FLAG_IPV6] => 2097152
            [FILTER_FLAG_NO_RES_RANGE] => 4194304
            [FILTER_FLAG_NO_PRIV_RANGE] => 8388608
            [FILTER_FLAG_GLOBAL_RANGE] => 268435456
            [FILTER_FLAG_HOSTNAME] => 1048576
            [FILTER_FLAG_EMAIL_UNICODE] => 1048576
      
```
